### PR TITLE
Add support for asymmetric input/output dimensions in Entangler Layer

### PR DESCRIPTION
**Pull request by jacksonwb**
* Added support for asymmetric input/output dimensions
* Changed entangler layer to create complete node levels

### CODE REVIEW

File: tensornetwork/tn_keras/test_layer.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ def dummy_data(request):
 
 @pytest.fixture(params=[
     'DenseDecomp', 'DenseMPO', 'DenseCondenser', 'DenseExpander',
-    'DenseEntangler'
+    'DenseEntangler', 'DenseEntanglerOut'
```
Comment (Thenerdstation) [inline tensornetwork/tn_keras/test_layer.py:28]: What is this?

Comment (jacksonwb) [reply tensornetwork/tn_keras/test_layer.py:28]: Thats an entangler model with a different output size to input size. Meant to change that to a more informative name.

File: tensornetwork/tn_keras/test_layer.py [hunk 2]
```diff
@@ -333,7 +348,35 @@ def test_entangler_num_parameters(dummy_data):
...
+  np.testing.assert_equal(expected_num_parameters, model.count_params())
+
+
+def test_entangler_asymmetric_num_parameters(dummy_data):
+  # Disable the redefined-outer-name violation in this function
+  # pylint: disable=redefined-outer-name
+  data, _ = dummy_data
+
+  num_legs = 3
+  num_levels = 3
```
Comment (Thenerdstation) [inline tensornetwork/tn_keras/test_layer.py:363]: can you add tests that vary these parameters?

File: tensornetwork/tn_keras/entangler.py [hunk 3]
```diff
@@ -102,24 +102,32 @@ def is_perfect_root(n, n_nodes):
...
+
     self.leg_dim = round(input_shape[-1]**(1. / self.num_legs))
-    self.num_nodes = self.num_levels
+    self.out_leg_dim = round(self.output_dim**(1. / self.num_legs))
+    self.num_nodes = self.num_levels * (self.num_legs - 1)
 
     for i in range(self.num_nodes):
+      current_level = i // (self.num_legs - 1)
+      node_out_dim = self.out_leg_dim
+      node_connect_dim = self.out_leg_dim
```
Comment (bpenchas) [inline tensornetwork/tn_keras/entangler.py:121]: I think we should either (a) set this to the min of `out_leg_dim` and `in_leg_dim`, so we use the smaller of the two options or (b) expose a separate `bond_dim` parameter the user can tune and supply some guidance around reasonable values. (a) seems more clear for end users, while (b) seems better in a research context (where we are likely to want to try different `bond_dim` values and check performance.

File: tensornetwork/tn_keras/entangler.py [hunk 4]
```diff
@@ -102,24 +102,32 @@ def is_perfect_root(n, n_nodes):
...
     ), f'Input dim {input_shape[-1]}**(1. / {self.num_legs}) must be round.'
 
+    assert (
+        is_perfect_root(self.output_dim, self.num_legs)
+    ), f'Output dim {self.output_dim}**(1. / {self.num_legs}) must be round.'
+
     self.leg_dim = round(input_shape[-1]**(1. / self.num_legs))
-    self.num_nodes = self.num_levels
+    self.out_leg_dim = round(self.output_dim**(1. / self.num_legs))
+    self.num_nodes = self.num_levels * (self.num_legs - 1)
```
Comment (bpenchas) [inline tensornetwork/tn_keras/entangler.py:116]: Can you add a note to the docstring that makes it clear `num_levels` represents *complete* levels i.e. more than 1 node per level?

File: tensornetwork/tn_keras/test_layer.py [hunk 5]
```diff
@@ -333,8 +348,45 @@ def test_entangler_num_parameters(dummy_data):
...
+  if leg_dim > out_leg_dim:
+    primary, secondary = secondary, primary
+
+  expected_num_parameters = (num_levels - 1) * (num_legs - 1) * (primary**4) + (
+      num_legs - 2) * primary**3 * secondary + primary**2 * secondary**2 + (
+          out_leg_dim**num_legs)
+
 
+  data = np.random.randint(10, size=(10, data_shape[0]))
+  model(data)
```
Comment (bpenchas) [inline tensornetwork/tn_keras/test_layer.py:389]: Can you check that the output shapes are the same as the expected output shapes for all of the permutations of num_levels, num_legs, leg_dims?

**Comment by Thenerdstation** [comment]
Ben you have write access so feel free to merge this once the tests pass.

**Pull request merged**